### PR TITLE
WT-4310 Add option to not abort in diagnostic builds on data corruption

### PR DESCRIPTION
**Pull request by sulabhM**

### CODE REVIEW

**Comment by sulabhM** [comment]
@agorrod , what do you think about this approach. I am still thinking about how to test this, suggestions welcome.

**Comment by sulabhM** [comment]
@sueloverso , @agorrod , I am not sure if this is in the correct direction. I have left a comment on the Jira ticket explaining my doubts.

File: dist/api_data.py [hunk 1]
```diff
@@ -493,6 +493,9 @@ def __ge__(self, other):
     Config('debug_mode', '', r'''
         control the settings of various extended debugging features''',
         type='category', subconfig=[
+        Config('abort_on_corruption_in_diagnostic', 'true', r'''
```
Comment (sueloverso) [inline dist/api_data.py:496]: That is really long and trying to remember all the words in the right order is hard. How about something like `corruption_abort` instead?

File: src/include/connection.h [hunk 2]
```diff
@@ -455,10 +455,11 @@ struct __wt_connection_impl {
     uint32_t debug_log_cnt;  /* Log file retention count */
 
 /* AUTOMATIC FLAG VALUE GENERATION START */
-#define WT_CONN_DEBUG_CKPT_RETAIN 0x1u
-#define WT_CONN_DEBUG_CURSOR_COPY 0x2u
-#define WT_CONN_DEBUG_REALLOC_EXACT 0x4u
-#define WT_CONN_DEBUG_SLOW_CKPT 0x8u
+#define WT_CONN_DEBUG_CKPT_RETAIN 0x01u
+#define WT_CONN_DEBUG_CURSOR_COPY 0x02u
+#define WT_CONN_DEBUG_DATA_CORRUPTION_ABORT_DIAG 0x04u
```
Comment (sueloverso) [inline src/include/connection.h:460]: Similarly this could be named `WT_CONN_DEBUG_CORRUPTION_ABORT`.

**Review by sueloverso — CHANGES_REQUESTED** [review]
As I mentioned in the ticket, I think there are two tests that could be modified to use this and test it out as well as perform their test tasks now without having to create a subprocess.

**Comment by agorrod** [comment]
The approach seems OK to me. I've pinged Dan in the JIRA ticket - I think he'll have the most relevant opinion.
I agree that we should be using it internally somewhere before going further.

**Comment by sulabhM** [comment]
Thanks @sueloverso , I am looking at those python tests

**Comment by sulabhM** [comment]
@sueloverso , I still have 3 cases that fail with txn19. Would you mind taking a look?

There have been a few cases that are known to produce an error other than "WT_TRY_SALVAGE: database corruption detected", and I have updated the test to take care of them. I am not sure of the remaining errors.

Two of them are in log file corruption where log file read errors out with "failed to read" or "no such file", can you confirm that we will not return salvage in these cases.

Other two cases are failures in tear down or raising assertion with the removal of turtle file and WiredTiger file respectively.

**Comment by sulabhM** [comment]
@sueloverso , please have another look. I had to modify txn19.py to account for extra messages that we need to process if not using "wt" utility. It is also worthwhile to review these cases as they do not return a salvage error back to the application.

**Comment by sulabhM** [comment]
test_txn19.py is passing locally. Looks like a different output is getting generated on the stdout on evergreen.

File: src/support/err.c [hunk 3]
```diff
@@ -369,23 +372,20 @@ __wt_panic_func(WT_SESSION_IMPL *session, int error, const char *func, int line,
...
-/*
- * Confusing #ifdef structure because gcc/clang knows the abort call won't return, and Visual Studio
- * doesn't.
- */
 #if defined(HAVE_DIAGNOSTIC)
-    __wt_abort(session); /* Drop core if testing. */
-                         /* NOTREACHED */
+    /* Drop core if testing, unless it is data corruption and we have been asked not to panic. */
+    if (!F_ISSET(conn, WT_CONN_DATA_CORRUPTION) ||
+      FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_CORRUPTION_ABORT))
```
Comment (sueloverso) [inline src/support/err.c:378]: Minor: I suggest rewriting the comment so it is the same logic as the code.

**Review by sueloverso — APPROVED** [review]
I think this is the right direction. I only have one minor comment.

**Comment by sueloverso** [comment]
@ddanderson can you take a look at this, especially per `test_txn19.py`? You worked on the original log corruption code and wrote the initial version of `test_txn19.py` (I think?).

File: test/suite/test_txn19.py [hunk 4]
```diff
@@ -286,30 +287,24 @@ def test_corrupt_log(self):
...
-        # Use -R to force recover=on, which is the default for
-        # wiredtiger_open, (wt utilities normally have recover=error)
-        self.runWt(['-h', newdir, '-C', self.base_config, '-R', 'list'],
-            errfilename=errfile, outfilename=outfile, failure=expect_fail,
-            closeconn=False)
-
         if expect_fail:
-            self.check_file_contains(errfile,
-                'WT_TRY_SALVAGE: database corruption detected')
+            with self.expectedStdoutPattern('.'):
```
Comment (ddanderson) [inline test/suite/test_txn19.py:293]: If feasible, we should have something specific to match in the pattern.  You can match multiple outputs, like
```
with self.expectedStdoutPattern('(this|that)'):
```

**Review by ddanderson — APPROVED** [review]
I suggested a possible improvement, otherwise LGTM.

File: src/support/err.c (hunk 3 above)
Comment (sulabhM) [reply src/support/err.c:385]: done. I added a more elaborate comment on what we are trying to do.

File: test/suite/test_txn19.py (hunk 4 above)
Comment (sulabhM) [reply test/suite/test_txn19.py]: done.

**Comment by sulabhM** [comment]
autobuild: test this please

**Comment by sulabhM** [comment]
evergreen retry

**Pull request merged**